### PR TITLE
Feature/layering

### PR DESCRIPTION
**Pull request by fmontei**
This PR adds the following functionality to Deckhand:
- Add layering logic to Deckhand's engine, allowing documents to be layered together, as per
the requirements in the design document.
- Add support for merge, delete and replace actions.

New unit tests were also added:
- Add unit tests to test basic merge actions.
- Add unit tests to handle more complicate layering scenarios.
- Add unit tests to handle basic scenarios included in design document.

### CODE REVIEW

File: deckhand/tests/unit/engine/test_document_layering.py
Comment (mark-burnett) [inline deckhand/tests/unit/engine/test_document_layering.py]: I think my expectations diverge a bit from this test.  @bryan-strassner and I have discussed having an implicit action of `{path: ., method: merge}` at the beginning of every layering definition, so that by default we include all the data from the child document, but essentially I would only expect to see the child data that's explicitly called out to be included in the render.

Case `path: .` is as I expect.

Case `path: .a` I would expect  `{a: {x: 7, y: 2, z: 3}, c: 9}`

Case `path: .b` I would expect `{a: {x: 1, y: 2}, b: 4, c: 9}`

Case `path: .c` I would expect `{a: {x: 1, y: 2}, c: 9}` -- Actually, maybe I would expect a validation error, since there is no `c` key in the child.

If we don't add the implicit merge as the first action, then I think we should consider a validation that looks for data in the child document that would not be included.  I think it's not very intuitive to have data in the child document that doesn't make it in, so maybe we should add the implicit action.  Thoughts?

Comment (fmontei) [reply deckhand/tests/unit/engine/test_document_layering.py]: @mark-burnett  I rewrote my unit tests to make it easier to keep track of changes, so for:

```python

            kwargs = {
                "_GLOBAL_DATA_": {"data": {"a": {"x": 1, "y": 2}, "c": 9}},
                "_SITE_DATA_": {"data": {"a": {"x": 7, "z": 3}, "b": 4}},
                "_SITE_ACTIONS_": {
                    "actions": [{"method": "merge", "path": ".a"}]}
            }
```

Does this translate to:

1) Keep all the site data but merge the site data in path ".a" with the global data in path ".a". Or:
2) Only keep the result of merging the site data in path ".a" with the global data in path ".a".

The second is not self-evident because "merge" implies "this plus this" versus "this plus this...except this".

Want to clarify this before making changes everywhere.

Comment (fmontei) [reply deckhand/tests/unit/engine/test_document_layering.py]: Also, why is "Case `path: .a` I would expect `{a: {x: 7, y: 2, z: 3}, c: 9}`" true? Why is "c" included in the result? Why not "b"? If the merge path is ".a" why is "b" being dropped from the site data and "c" being included from the global data?

I'd imagine that should only be possible if:

 ```python

    kwargs = {
                "_GLOBAL_DATA_": {"data": {"a": {"x": 1, "y": 2}, "c": 9}},
                "_SITE_DATA_": {"data": {"a": {"x": 7, "z": 3}, "b": 4}},
                "_SITE_ACTIONS_": {
                    "actions": [{"method": "merge", "path": ".a"}, {"method": "merge", "path": ".c"}, {"method": "delete", "path": ".b"}]}
            }
```

But as you say above, this would be impossible, since if ".c" is not in the child, you expect a validation error. So how do you get `"c": 9` into the child other than doing a "merge" with a path of `"."`?

**Review by alanmeadows — APPROVED** [review]
Felipe, this is an impressive set of Python work.

File: deckhand/engine/document.py
Comment (fmontei) [inline deckhand/engine/document.py]: Better to call this ``to_dict()``.

File: deckhand/engine/layering.py
Comment (fmontei) [inline deckhand/engine/layering.py]: The "Use copy..." comment is no longer valid and should be removed.

Comment (fmontei) [reply deckhand/engine/layering.py]: Actually, it should be moved to L125 below.

File: deckhand/engine/layering.py
Comment (fmontei) [inline deckhand/engine/layering.py]: This'll be a little easier to understand if it's ``d.to_dict() for d in self.layered_docs``.

File: deckhand/engine/layering.py
Comment (fmontei) [inline deckhand/engine/layering.py]: Add TODO saying this should defer to schema validation. Validation should've been done by now,

File: deckhand/engine/layering.py
Comment (fmontei) [inline deckhand/engine/layering.py]: This comment is deprecated. "Parent" is no longer added -- rather "children" is.

File: deckhand/engine/layering.py
Comment (fmontei) [inline deckhand/engine/layering.py]: Add comment: "all_children is counter utility for verifying that each document has exactly one parent."

**Comment by fmontei** [comment]
@alanmeadows Thanks. Since you approved, just updated some comments before merge. Just need @mark-burnett to look at this to make sure the logic is correct when he's back from vacation.

**Comment by fmontei** [comment]
Moved to Gerrithub: https://review.gerrithub.io/#/c/373140/

**Pull request closed without merging**
